firestore/src/android/field_value_android.cc

```diff
@@ -389,15 +406,24 @@ bool operator==(const FieldValueInternal& lhs, const FieldValueInternal& rhs) {
 }
 
 template <typename T>
-T FieldValueInternal::Cast(jni::Env& env, Type type) const {
+Local<T> FieldValueInternal::Cast(jni::Env& env, Type type) const {
+  if (cached_type_ == Type::kNull) {
+    FIREBASE_ASSERT(env.IsInstanceOf(object_.get(env), T::GetClass()));
+    cached_type_ = type;
+  } else {
+    FIREBASE_ASSERT(cached_type_ == type);
+  }
+  return object_.get(env).CastTo<T>();
+}
+
+Local<String> FieldValueInternal::CastString(jni::Env& env, Type type) const {
```

Comment:
You can get rid of this special `CastString` method. Here is what worked for me:

```
template <typename T>
T FieldValueInternal::Cast(jni::Env& env, Type type) const {
  if (cached_type_ == Type::kNull) {
    FIREBASE_ASSERT(env.IsInstanceOf(object_, T::GetClass()));
    cached_type_ = type;
  } else {
    FIREBASE_ASSERT(cached_type_ == type);
  }
  auto untyped_value = object_.get(env);
  auto typed_value = static_cast<jni::JniType<T>>(untyped_value.get());
  return T(typed_value);
}
```

Note that this assumes that you overrided Env::IsInstanceOf to take ArenaRef.

Reply:
Hmm, on second thought this might crash. If it does crash, then instead of creating a "CastString" method, create a specialization of the Cast template method for String.

Reply:
I don't think the above solution would work since typed_value contains a local reference jobject and it would be deleted when the function returns.

For template specialization, I am not sure what it would look like. `CastString` doesn't have different parameters compare to `Cast`. It can get rid of the second parameter but still need to have `String` inside the name for the readability.

Reply:
A template specialization would look something like this:

```cpp
template <>
Local<String> FieldValueInternal::Cast<String>(jni::Env& env, Type type) const {
  ...
}
```